File: src/main/java/com/openlattice/conductor/rpc/ConductorElasticsearchCall.java
/*
 * Copyright (C) 2018. OpenLattice, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * You can contact the owner of the copyright at support@openlattice.com
 *
 */

package com.openlattice.conductor.rpc;

import java.util.UUID;
import java.util.concurrent.Callable;
import java.util.function.Function;

import com.google.common.base.Preconditions;

public class ConductorElasticsearchCall<T> implements Callable<T> {
    private final UUID                                   userId;
    private final Function<ConductorElasticsearchApi, T> f;
    private final ConductorElasticsearchApi              api;

    public ConductorElasticsearchCall(
            UUID userId,
            Function<ConductorElasticsearchApi, T> call,
            ConductorElasticsearchApi api ) {
        this.userId = Preconditions.checkNotNull( userId );
        this.f = Preconditions.checkNotNull( call );
        this.api = api;
    }

    public UUID getUserId() {
        return userId;
    }

    @Override
    public T call() throws Exception {
        return f.apply( api );
    }

    public static <T> ConductorElasticsearchCall<T> wrap( Function<ConductorElasticsearchApi, T> f ) {
        return new ConductorElasticsearchCall<T>( UUID.randomUUID(), f, null );
    }

    public Function<ConductorElasticsearchApi, T> getFunction() {
        return f;
    }
}

File: src/main/kotlin/com/openlattice/conductor/rpc/ConductorElasticsearchCall.kt
/*
 * Copyright (C) 2019. OpenLattice, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * You can contact the owner of the copyright at support@openlattice.com
 *
 *
 */
package com.openlattice.conductor.rpc

import java.util.UUID
import java.util.concurrent.Callable
import java.util.function.Function

data class ConductorElasticsearchCall<T>(
        val userId: UUID,
        val function: Function<ConductorElasticsearchApi, T>,
        private val api: ConductorElasticsearchApi?
) : Callable<T> {
    companion object {
        @JvmStatic
        fun <T> wrap(f: Function<ConductorElasticsearchApi, T>): ConductorElasticsearchCall<T> {
            return ConductorElasticsearchCall(UUID.randomUUID(), f, null)
        }
    }

    override fun call(): T {
        return function.apply(api!!)
    }
}